Dockerfile for Identity added. Errors on build custom exceptions

YaroslavKSE opened the pull request:
Dockerfile for Identity added. Added docker-compose for starting all the dockerfiles. I have got errors on building custom exceptions. @Bardin08 can you help me to fix them? 

See details here:
<details><summary>Details</summary>
<p>

S D:\C#\AcademicHub> docker-compose up -d
[+] Building 18.3s (21/23)
 => [internal] load build definition from Dockerfile                                                                                                                                            0.0s
 => => transferring dockerfile: 1.17kB                                                                                                                                                          0.0s
 => [internal] load .dockerignore                                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                                 0.0s
 => [internal] load metadata for mcr.microsoft.com/dotnet/sdk:7.0                                                                                                                               0.0s
 => [internal] load metadata for mcr.microsoft.com/dotnet/aspnet:7.0                                                                                                                            0.0s
 => [base 1/2] FROM mcr.microsoft.com/dotnet/aspnet:7.0                                                                                                                                         0.0s
 => [internal] load build context                                                                                                                                                               0.8s
 => => transferring context: 939B                                                                                                                                                               0.8s
 => [build  1/13] FROM mcr.microsoft.com/dotnet/sdk:7.0                                                                                                                                         0.0s
 => CACHED [base 2/2] WORKDIR /app                                                                                                                                                              0.0s
 => CACHED [final 1/2] WORKDIR /app                                                                                                                                                             0.0s
 => CACHED [build  2/13] WORKDIR /src                                                                                                                                                           0.0s
 => CACHED [build  3/13] COPY [Identity/Identity/Identity.csproj, Identity/]                                                                                                                    0.0s
 => CACHED [build  4/13] COPY [Identity/Identity.Data/Identity.Data.csproj, Identity.Data/]                                                                                                     0.0s
 => CACHED [build  5/13] COPY [Identity/Identity.Core/Identity.Core.csproj, Identity.Core/]                                                                                                     0.0s
 => [build  6/13] COPY [AcademicHub.Common/CustomExceptions/CustomExceptions.csproj, AcademicHub.Common/CustomExceptions/]                                                                      0.0s
 => [build  7/13] RUN dotnet restore "Identity/Identity.csproj"                                                                                                                                12.7s
 => [build  8/13] COPY [Identity/Identity/*.cs, Identity/Identity/]                                                                                                                             0.0s
 => [build  9/13] COPY [Identity/Identity.Data/*.cs, Identity/Identity.Data/]                                                                                                                   0.0s
 => [build 10/13] COPY [Identity/Identity.Core/*.cs, Identity/Identity.Core/]                                                                                                                   0.0s
 => [build 11/13] COPY [AcademicHub.Common/CustomExceptions/*.cs, AcademicHub.Common/CustomExceptions/]                                                                                         0.0s
 => [build 12/13] WORKDIR /src/Identity                                                                                                                                                         0.0s
 => ERROR [build 13/13] RUN dotnet build "Identity.csproj" -c Release -o /app/build                                                                                                             4.4s
------
 > [build 13/13] RUN dotnet build "Identity.csproj" -c Release -o /app/build:
#0 0.592 MSBuild version 17.7.4+3ebbd7c49 for .NET
#0 1.178   Determining projects to restore...
#0 1.200   Skipping project "/src/Identity.Core/AcademicHub.Common/CustomExceptions/CustomExceptions.csproj" because it was not found.
#0 1.249   Skipping project "/src/Identity.Core/AcademicHub.Common/CustomExceptions/CustomExceptions.csproj" because it was not found.
#0 1.621   All projects are up-to-date for restore.
#0 3.664   Identity.Data -> /app/build/Identity.Data.dll
#0 3.680 /usr/share/dotnet/sdk/7.0.404/Microsoft.Common.CurrentVersion.targets(2128,5): warning : The referenced project 'AcademicHub.Common/CustomExceptions/CustomExceptions.csproj' does not exist
. [/src/Identity.Core/Identity.Core.csproj]
#0 3.909   Identity.Core -> /app/build/Identity.Core.dll
#0 4.340 /src/Identity/Identity/Program.cs(1,7): error CS0246: The type or namespace name 'CustomExceptions' could not be found (are you missing a using directive or an assembly reference?) [/src/I
dentity/Identity.csproj]
#0 4.357
#0 4.357 Build FAILED.
#0 4.357
#0 4.357 /usr/share/dotnet/sdk/7.0.404/Microsoft.Common.CurrentVersion.targets(2128,5): warning : The referenced project 'AcademicHub.Common/CustomExceptions/CustomExceptions.csproj' does not exist
. [/src/Identity.Core/Identity.Core.csproj]
#0 4.357 /src/Identity/Identity/Program.cs(1,7): error CS0246: The type or namespace name 'CustomExceptions' could not be found (are you missing a using directive or an assembly reference?) [/src/I
dentity/Identity.csproj]
#0 4.357     1 Warning(s)
#0 4.357     1 Error(s)
#0 4.357
#0 4.358 Time Elapsed 00:00:03.63


</p>
</details>

Bardin08 commented on Identity/Identity/Dockerfile:
Unfortunately, this approach won't work this time.
Follow these steps to containerize this & other services:
1. At the repo's root directory, create `Dockerfile.$service_name`
2. Use this docker file and replace placeholders. Placeholders starts with `$`
```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:$dotnet_version-alpine AS build-env
WORKDIR src

COPY ./$service ./$service
COPY ./AcademicHub.Common ./AcademicHub.Common
RUN cd $service && dotnet restore
RUN cd $service && dotnet publish -c Release -o out

FROM mcr.microsoft.com/dotnet/aspnet:$dotnet_version-alpine
COPY --from=build-env /src/$service/out ./
ENTRYPOINT ["dotnet", "./$service.dll"]
```
3. Let me know if plays for u

YaroslavKSE replied:
Unfortunatelly, this solution didn't help I have got the same error as before. 
<details><summary>Details</summary>
<p>

 docker-compose up -d
[+] Building 9.7s (18/19)
 => [internal] load build definition from Dockerfile.Identity                             0.0s
 => => transferring dockerfile: 1.04kB                                                    0.0s
 => [internal] load .dockerignore                                                         0.0s
 => => transferring context: 2B                                                           0.0s
 => [internal] load metadata for mcr.microsoft.com/dotnet/aspnet:7.0-alpine               0.1s
 => [internal] load metadata for mcr.microsoft.com/dotnet/sdk:7.0-alpine                  0.1s
 => [internal] load build context                                                         0.1s
 => => transferring context: 47.23kB                                                      0.0s
 => [build-env  1/11] FROM mcr.microsoft.com/dotnet/sdk:7.0-alpine@sha256:4087b301a58dfb  0.0s
 => [stage-1 1/3] FROM mcr.microsoft.com/dotnet/aspnet:7.0-alpine@sha256:ba78f81014ab4a5  0.0s
 => CACHED [stage-1 2/3] WORKDIR /app                                                     0.0s
 => CACHED [build-env  2/11] WORKDIR /src                                                 0.0s
 => CACHED [build-env  3/11] COPY [Identity/Identity/Identity.csproj, Identity/]          0.0s
 => [build-env  4/11] COPY [Identity/Identity.Core/Identity.Core.csproj, Identity.Core/]  0.0s
 => [build-env  5/11] COPY [AcademicHub.Common/CustomExceptions/CustomExceptions.csproj,  0.1s
 => [build-env  6/11] RUN dotnet restore "Identity/Identity.csproj"                       3.7s
 => [build-env  7/11] RUN dotnet restore "Identity.Core/Identity.Core.csproj"             1.6s
 => [build-env  8/11] COPY [Identity/, Identity/]                                         0.1s
 => [build-env  9/11] COPY [Identity/Identity.Core/, Identity.Core/]                      0.0s
 => [build-env 10/11] COPY [AcademicHub.Common/, AcademicHub.Common/]                     0.0s
 => ERROR [build-env 11/11] RUN dotnet publish "Identity/Identity.csproj" -c Release -o   3.8s
------
 > [build-env 11/11] RUN dotnet publish "Identity/Identity.csproj" -c Release -o out:
#0 0.670 MSBuild version 17.7.4+3ebbd7c49 for .NET
#0 1.508   Determining projects to restore...
#0 1.537   Skipping project "/src/Identity.Core/AcademicHub.Common/CustomExceptions/CustomExcep
tions.csproj" because it was not found.
#0 1.537   Skipping project "/src/Identity.Data/Identity.Data.csproj" because it was not found.
#0 1.544   Skipping project "/src/Identity.Core/AcademicHub.Common/CustomExceptions/CustomExcep
tions.csproj" because it was not found.
#0 1.544   Skipping project "/src/Identity.Data/Identity.Data.csproj" because it was not found.
#0 2.221   Restored /src/Identity.Core/Identity.Core.csproj (in 199 ms).
#0 2.226   1 of 2 projects are up-to-date for restore.
#0 2.448 /usr/share/dotnet/sdk/7.0.404/Microsoft.Common.CurrentVersion.targets(2128,5): warning
 : The referenced project 'AcademicHub.Common/CustomExceptions/CustomExceptions.csproj' does not exist. [/src/Identity.Core/Identity.Core.csproj]
#0 2.448 /usr/share/dotnet/sdk/7.0.404/Microsoft.Common.CurrentVersion.targets(2128,5): warning : The referenced project '../Identity.Data/Identity.Data.csproj' does not exist. [/src/Identity.Core/
Identity.Core.csproj]
#0 3.751 /src/Identity.Core/UsersService.cs(1,16): error CS0234: The type or namespace name 'Data' does not exist in the namespace 'Identity' (are you missing an assembly reference?) [/src/Identity
.Core/Identity.Core.csproj]
#0 3.751 /src/Identity.Core/UsersService.cs(14,20): error CS0246: The type or namespace name 'IUsersRepository' could not be found (are you missing a using directive or an assembly reference?) [/sr
c/Identity.Core/Identity.Core.csproj]
#0 3.751 /src/Identity.Core/UsersService.cs(21,23): error CS0246: The type or namespace name 'IUsersRepository' could not be found (are you missing a using directive or an assembly reference?) [/sr
c/Identity.Core/Identity.Core.csproj]
------
failed to solve: executor failed running [/bin/sh -c dotnet publish "Identity/Identity.csproj" -c Release -o out]: exit code: 1
PS D:\C#\AcademicHub>


</p>
</details>

Bardin08 commented on Identity/Identity/Identity.csproj:
```diff
@@ -5,6 +5,7 @@
         <Nullable>enable</Nullable>
         <ImplicitUsings>enable</ImplicitUsings>
         <UserSecretsId>7605f2a2-354b-4ce1-bb65-077ad4e18251</UserSecretsId>
+        <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
```
Drop this line

Bardin08 commented on docker-compose.yml:
```diff
@@ -0,0 +1,7 @@
+version: '3.8'
```
Let's add DB here (if applicable) and define a separated network for each service.

Ex:
```yml
services:
    postgres:
        image: bitnami/postgresql:latest
        container_name: identity-postgres-container
        environment:
            - POSTGRESQL_DATABASE=Identity
            - POSTGRESQL_USER=identityuser
            - POSTGRESQL_PASSWORD=identityuser
        ports:
            - "5440:5432"
        networks:
            - academic_hub_backend_vpn

    identity:
        build:
            context: .
            dockerfile: Dockerfile.Identity
        restart: no
        container_name: academic_hub_identity
        environment:
            - NAME=VALUE
        ports:
            - $local:$service
        networks:
            - academic_hub_backend_vpn

networks:
    academic_hub_backend_vpn:
        name: academic_hub_backend_vpn
```

Bardin08 replied:
then show me the error

YaroslavKSE commented on Identity/Identity.Core/Identity.Core.csproj:
```diff
@@ -7,8 +7,10 @@
     </PropertyGroup>
 
     <ItemGroup>
-      <ProjectReference Include="..\..\AcademicHub.Common\CustomExceptions\CustomExceptions.csproj" />
```
Done

YaroslavKSE commented:
> Dockerfile for Identity added. Added docker-compose for starting all the dockerfiles. I have got errors on building custom exceptions. @Bardin08 can you help me to fix them?
> 
> See details here:
> 
> Details

Vladyslav, I have found the similiar issue to our with connection to db through docker. The guy who fixed it says that: "Finally, I was able to resolve my own problem and it wasn't in the docker-compose.yml file at all. Somewhere in the application code, connection string was changed to look for localhost as a host instead of postgres.
After changing it back to postgres, everything was fine."
I have tried to fix connection string, however it seems that the problem is deeper. Can you help me to fix that?
https://stackoverflow.com/questions/61535772/what-am-i-doing-wrong-in-docker-compose-for-netcore-and-postgres

YaroslavKSE commented:
> Dockerfile for Identity added. Added docker-compose for starting all the dockerfiles. I have got errors on building custom exceptions. @Bardin08 can you help me to fix them?
> 
> See details here:
> 
> Details

Fixed

Bardin08 commented on CourseStream/CourseStream.Core/CourseStreamService.cs:
```diff
@@ -23,10 +23,6 @@ public class CourseStreamAddDto
 public class CourseStreamService
 {
 	private readonly ICourseStreamRepository _repository;
-	public CourseStreamService()
```
how this change affect local development?

Bardin08 replied:
I believe this can cause issues with applying migrations

YaroslavKSE replied:
This should be deleted in order to launch correctly application in docker. All configuration are done in startup

Bardin08 commented on CourseStream/CourseStream.Data/CourseStreamDBContext.cs:
```diff
@@ -1,22 +1,13 @@
-public class CourseStreamDBContext : DbContext
+public class CourseStreamDbContext : DbContext
 {
-    public DbSet<CourseStream> CourseStreams { get; set; }
-    public DbSet<Enrollment> Enrollments { get; set; }
-    public CourseStreamDBContext()
-    {
-        
-    }
-    public CourseStreamDBContext(DbContextOptions<CourseStreamDBContext> options) : base(options)
-    {
-    }
-
-    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
```
Same as prev. How this will affect on our current migrations applying flow?

YaroslavKSE replied:
This should be deleted in order to launch correctly application in docker. All configuration are done in startup

Bardin08 commented on CourseStream/CourseStream.Data/Migrations/20231113164931_InitialMigration.Designer.cs:
```diff
@@ -10,7 +10,7 @@
 {
-    [DbContext(typeof(CourseStreamDBContext))]
+    [DbContext(typeof(CourseStreamDbContext))]
     [Migration("20231113164931_InitialMigration")]
```
Looks a bit destructive. Won't this impact on existing DBs with data?

YaroslavKSE replied:
There was a problem after merging main into Docker Support branch. I needed to change the naming because migration won't be applied

Bardin08 commented on CourseStream/CourseStream.UnitTests/CourseStreamServiceTest.cs:
```diff
@@ -126,7 +126,8 @@ public async Task ListAsync_ReturnsMappedCourseStreams()
         {
             // Arrange & Act
-            var service = new CourseStreamService();
+            var mockRepo = new Mock<ICourseStreamRepository>();
```
Define mocks at the class level. Not at the methods. Each method is an isolated test and mocks have to be cleaned up by the framework (xUnit). Here confirmation required. I haven't write tests for quiet a long and can miss some important aspects

Bardin08 commented on CourseStream/CourseStream/Program.cs:
```diff
@@ -1,22 +1,29 @@
+// Configure your DbContext
+builder.Services.AddDbContext<CourseStreamDbContext>(options =>
+	options.UseNpgsql(connectionString));
+builder.Services.AddScoped<CourseStreamService>();
```
Maybe it makes sense to create an issue to create an interface for `CourseStreamService`?

Bardin08 replied:
Please, do not fix this in scope of this PR

YaroslavKSE replied:
I needed to do this, because in other way dockers didn't start

Bardin08 commented on CourseStream/CourseStream/Program.cs:
```diff
@@ -48,4 +55,13 @@
 app.MapPost("/courseStreams/add", EndpointHandlers.AddCourse);
 app.MapPost("courseStreams/enroll", EndpointHandlers.EnrollStudent);
+
+// Apply EF Core Migrations
+
+using (var scope = app.Services.CreateScope())
+{
+	var dbContext = scope.ServiceProvider.GetRequiredService<CourseStreamDbContext>();
+	dbContext.Database.Migrate();
+}
```
Better to move to a separated class/method to keep entrypoint tiny

YaroslavKSE replied:
Okay, I'll think about it, for now let's leave this working solution

Bardin08 commented on CourseTemplate/CourseTemplate.Core/CourseTemplateService.cs:
```diff
@@ -17,11 +17,6 @@ public class ViewCourseTemplateDto
     {
         private readonly ICourseTemplateRepository _repository;
-
-        public CourseTemplateService()
-        {
-            _repository = new CourseTemplateRepository();
-        }
```
How this impact on migrations applying?

YaroslavKSE replied:
This should be deleted in order to launch correctly application in docker. All configuration are done in startup

Bardin08 commented on CourseTemplate/CourseTemplate.Data/CorseTemplateRepository.cs:
```diff
@@ -8,12 +8,7 @@ public class CourseTemplateRepository : ICourseTemplateRepository
 
 	public CourseTemplateRepository(CourseTemplateDbContext context)
 	{
-		_context = context ?? throw new ArgumentNullException(nameof(context));
-	}
-
-	public CourseTemplateRepository()
-	{
-		_context = new CourseTemplateDbContext();
```
How this impact on migrations applying?

YaroslavKSE replied:
This should be deleted in order to launch correctly application in docker. All configuration are done in startup

Bardin08 commented on CourseTemplate/CourseTemplate.Data/CourseTemplateDbContext.cs:
```diff
@@ -4,17 +4,12 @@ namespace CourseTemplate.Data;
 public class CourseTemplateDbContext : DbContext
 {
-	public DbSet<CourseTemplate> CourseTemplates { get; set; }
-	public CourseTemplateDbContext()
```
How this impact on migrations applying?

YaroslavKSE replied:
This should be deleted in order to launch correctly application in docker. All configuration are done in startup

Bardin08 commented on CourseTemplate/CourseTemplate/Program.cs:
```diff
@@ -1,19 +1,36 @@
 builder.Services.AddSwaggerGen();
 builder.Services.AddLogging();
+
+// Configure Forwarded Headers Middleware to handle reverse proxy server scenarios
+// builder.Services.Configure<ForwardedHeadersOptions>(options =>
+// {
+//     options.ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto;
+// });
```
Why commented?

YaroslavKSE replied:
For future we can use proxy. So I left these comments in advance with an example of implementation

Bardin08 commented:
@YaroslavKSE I added a review for one service, but please according to it validate all ur changes

Bardin08 commented:
@YaroslavKSE please confirm that migrations applied successful to newly created container

Bardin08 reviewed (approved):
comment above

YaroslavKSE commented:
> @YaroslavKSE please confirm that migrations applied successful to newly created container. Add screenshots or small video

Okay, I'll link screenshots

YaroslavKSE commented:
> @YaroslavKSE please confirm that migrations applied successful to newly created container. Add screenshots or small video
Here is the logs of the "Identity" startup job

![image](https://github.com/AcademicHubOrg/AcademicHub/assets/94047397/e210d73e-bb33-44c4-9829-a68b80c7b2bf)

qualia4 reviewed (approved):
Nice changes. Lets deploy
